python/pyiceberg/catalog/rest.py

```diff
@@ -184,28 +190,40 @@ def __init__(
         """
         self.properties = properties
         self.uri = properties[URI]
+        self.session = requests.Session()
+        self._set_session_ssl_config()
 
         if credential := properties.get(CREDENTIAL):
             properties[TOKEN] = self._fetch_access_token(credential)
+
+        self._set_session_headers()
         super().__init__(name, **self._fetch_config(properties))
 
+    def _set_session_ssl_config(self):
+        """Sets the client side and server side SSL cert verification, if provided as properties."""
+        if ssl_config := self.properties.get(SSL):
+            if ssl_ca_bundle := ssl_config.get(CA_BUNDLE):
+                self.session.verify = ssl_ca_bundle
+            if ssl_client := ssl_config.get(CLIENT):
+                if all(k in ssl_client for k in (CERT, KEY)):
+                    self.session.cert = (ssl_client[CERT], ssl_client[KEY])
+                elif ssl_client_cert := ssl_client.get(CERT):
+                    self.session.cert = ssl_client_cert
+
     def _check_valid_namespace_identifier(self, identifier: Union[str, Identifier]) -> Identifier:
         """The identifier should have at least one element"""
         identifier_tuple = Catalog.identifier_to_tuple(identifier)
         if len(identifier_tuple) < 1:
             raise NoSuchNamespaceError(f"Empty namespace identifier: {identifier}")
         return identifier_tuple
 
-    @property
-    def headers(self) -> Properties:
-        headers = {
-            "Content-type": "application/json",
-            "X-Client-Version": ICEBERG_REST_SPEC_VERSION,
-            "User-Agent": f"PyIceberg/{__version__}",
-        }
+    def _set_session_headers(self):
```

Comment:
I like this a lot, much nicer than having to pass in the headers every time.

Reply:
> Thanks for working on this @dhruv-pratap! I like it a lot. Another approach we could also take is to have a function `_create_session() -> Session` that will set up the session. This would both set the SSL and the headers, and in the constructor, we would do:
> 
> ```
> self.session = self._create_session()
> ```
> 
> This way we have a single place where we set up the session, instead of two methods. WDYT?

I did think about that initially, but the `self._fetch_access_token(credential)` REST call poses a kind-of chicken-and-egg problem. You need the session SSL config to be configured to make that REST call, and then the session's header has to be enriched with the Auth Token. Let me try to give it a shot to see what that looks like and send an amendment.

Reply:
Suggested changes pushed in recent commit. Please review.

Reply:
@dhruv-pratapI didn't think of the chicken-and-egg problem. The current version isn't working:
```
RESTError: HttpMediaTypeNotSupportedException: Content type 'application/json' not supported
```
It already sets the `application/json` header, but we need to set the `application/x-www-form-urlencoded` header for the signer:
https://github.com/apache/iceberg/blob/master/python/pyiceberg/catalog/rest.py#L237

Reply:
Sorry, I didn't have an OAuth environment to integration test it against.

Changed the order in recent commit, now setting the HTTP headers after fetching the OAuth token. Also, enforcing the content-type check to application/x-www-form-urlencoded for OAuth calls in unit tests.

Reply:
No problem at all. I've just checked and it works 👍🏻